sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -281,3 +281,21 @@ def test_percentile_numeric_stability():
     assert_warns_message(UserWarning, msg, kbd.fit, X)
     assert_array_almost_equal(kbd.bin_edges_[0], bin_edges)
     assert_array_almost_equal(kbd.transform(X), Xt)
+
+
+@pytest.mark.parametrize("input_dtype", [np.float16, np.float32, np.float64])
+@pytest.mark.parametrize("output_dtype", [None, np.float32, np.float64])
+@pytest.mark.parametrize('encode', ['ordinal', 'onehot', 'onehot-dense'])
+def test_dtype(input_dtype, output_dtype, encode):
+    # test dtype parameter to properly cast output dtype (output_dtype if
+    # specified or consistent with input_dtype instead).
+    kbd = KBinsDiscretizer(n_bins=3, encode=encode, dtype=output_dtype)
+    kbd.fit(X)
+    X_input = np.array(X, dtype=input_dtype)
+    output_dtype_ = output_dtype if output_dtype is not None else X_input.dtype
+    # wrong input_dtype are cast in np.float64 by default so the expected
+    # output_dtype will be 64 bit.
+    if input_dtype == np.float16 and output_dtype is None:
+        output_dtype_ = np.float64
+    Xt = kbd.transform(X_input)
+    assert Xt.dtype == output_dtype_
```

Comment:
Can we also please add another check, evaluating that the output of `fit_transform` for dtype=np.float32 and dtype=np.float64 is the same, with a comment
```
# TODO this check is redundant with common checks and can be removed
# once #16290 is merged
```
Thanks!